exotations/solvers/exotica_ddp_solver/include/exotica_ddp_solver/abstract_ddp_solver.h

```diff
@@ -0,0 +1,219 @@
+    std::vector<Eigen::MatrixXd> K_gains_, k_gains_;  ///!< Control gains.
+
+    ///\brief Computes the control gains for a the trajectory in the associated
+    ///     DynamicTimeIndexedProblem.
+    virtual void BackwardPass() = 0;
+
+    ///\brief Forward simulates the dynamics using the gains computed in the
+    ///     last BackwardPass;
+    /// @param alpha The learning rate.
+    /// @param ref_trajectory The reference state trajectory.
+    /// @return The cost associated with the new control and state trajectory.
+    double ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory);
+
+    Initializer base_parameters_;
+};
+
+template <typename Initializer>
+void AbstractDDPSolver<Initializer>::Solve(Eigen::MatrixXd& solution)
+{
+    if (!prob_) ThrowNamed("Solver has not been initialized!");
+    Timer planning_timer, backward_pass_timer, line_search_timer;
+
+    const int T = prob_->get_T();
+    const int NX = prob_->get_num_positions();
+    const int NU = prob_->get_num_controls();
+    const int N = NX + prob_->get_num_velocities();
+
+    prob_->ResetCostEvolution(GetNumberOfMaxIterations() + 1);
+
+    // initialize Gain matrices
+    K_gains_.assign(T, Eigen::MatrixXd(NU, N));
+    k_gains_.assign(T, Eigen::VectorXd(NU, 1));
+
+    // all of the below are not pointers, since we want to copy over
+    //  solutions across iterations
+    Eigen::MatrixXd new_U, global_best_U;
+    solution.resize(T, NU);
+
+    if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Running DDP solver for max " << GetNumberOfMaxIterations() << " iterations");
+
+    double last_cost = 0, global_best_cost = 0;
+    int last_best_iteration = 0;
+
+    for (int iteration = 0; iteration < GetNumberOfMaxIterations(); ++iteration)
+    {
+        // Backwards pass computes the gains
+        backward_pass_timer.Reset();
+        BackwardPass();
+        if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Backward pass complete in " << backward_pass_timer.GetDuration());
+
+        line_search_timer.Reset();
+        // forward pass to compute new control trajectory
+        // TODO: Configure line-search space from xml
+        // TODO (Wolf): What you are doing is a forward line-search when we may try a
+        //    backtracking line search for a performance improvement later - this just as an aside.
+        const Eigen::VectorXd alpha_space = Eigen::VectorXd::LinSpaced(10, 0.1, 1.0);
+        double current_cost = 0, best_alpha = 0;
+
+        Eigen::MatrixXd ref_trajectory = prob_->get_X();
+        // perform a linear search to find the best rate
+        for (int ai = 0; ai < alpha_space.rows(); ++ai)
+        {
+            double alpha = alpha_space(ai);
+            double cost = ForwardPass(alpha, ref_trajectory);
+
+            if (ai == 0 || cost < current_cost)
+            {
+                current_cost = cost;
+                new_U = prob_->get_U();
+                best_alpha = alpha;
+            }
```

Comment:
I think you can break this loop early if the cost is increasing.

Reply:
See below